rlay-backend-neo4j/src/lib.rs

```diff
@@ -220,76 +220,170 @@ impl Neo4jBackend {
 
         let client = self.client().await?;
 
         let kind_name: &str = entity.kind().into();
         let entity_val = serde_json::to_value(FormatWeb3(entity.clone())).unwrap();
         let val = entity_val.as_object().unwrap();
-        let mut values = Vec::new();
         let mut relationships = Vec::new();
         {
-            let mut add_relationship_value = |cid, key, value| {
-                let rel_query = format!(
-                            "MATCH (n:RlayEntity {{ cid: \"{0}\"}}) MERGE (m:RlayEntity {{ cid: {2} }}) MERGE (n)-[r:{1}]->(m)",
-                            cid, key, value
-                        );
-                relationships.push(rel_query);
-            };
-
-            for (key, value) in val {
+            #[derive(Serialize, Deserialize, Debug)]
+            struct RelationshipQueryPart {
+                cid: String,
+                kind_name: String
+            }
+
+            // fetch the parts of the payload that should become relationships
+            // aka. those that are not self-referencing CIDs
+            for (key, value) in val.iter() {
                 if key == "cid" || key == "type" {
                     continue;
                 }
                 if (kind_name == "DataPropertyAssertion"
                     || kind_name == "NegativeDataPropertyAssertion")
                     && key == "target"
                 {
-                    values.push(format!("n.{0} = {1}", key, value));
                     continue;
                 }
                 if kind_name == "Annotation" && key == "value" {
-                    values.push(format!("n.{0} = {1}", key, value));
                     continue;
                 }
                 if let Value::Array(array_val) = value {
-                    for relationship_value in array_val {
-                        add_relationship_value(cid.clone(), key, relationship_value);
+                    for array_val_cid in array_val {
+                        relationships.push(RelationshipQueryPart{
+                            cid: array_val_cid.as_str().unwrap().to_string(),
+                            kind_name: key.clone()
+                        });
                     }
                     continue;
                 }
                 if let Value::String(_) = value {
-                    add_relationship_value(cid.clone(), key, value);
+                    relationships.push(RelationshipQueryPart{
+                        cid: value.as_str().unwrap().to_string(),
+                        kind_name: key.clone()
+                    });
                 }
             }
         }
 
-        let mut statement_query = format!(
-            "MERGE (n:RlayEntity {{cid: \"{1}\"}}) SET n:{0}",
-            kind_name, cid
+        let statement_query_main = format!("
+            MERGE (n:RlayEntity {{cid: $entity.cid }})
+            SET (CASE WHEN $entity.type = 'Annotation' THEN n END).value = $entity.value
```

Comment:
I don't think this works? If it does that would be a cool discovery :)

Reply:
Oh it actually does, really cool!!

Reply:
Yeah, super nifty.

I found it originally on a blog - but now I can't find the link anymore :/